blockchain/types/derivesha/mux.go

```diff
@@ -54,23 +59,36 @@ func init() {
 	}
 }
 
-func InitDeriveSha(chainConfig *params.ChainConfig) {
+func InitDeriveSha(chainConfig *params.ChainConfig, govEngine GovernanceEngine) {
 	config = chainConfig
+	gov = govEngine
 	types.DeriveSha = DeriveShaMux
 	types.EmptyRootHash = EmptyRootHashMux
+	logger.Info("InitDeriveSha", "initial", config.DeriveShaImpl, "withGov", gov != nil)
 }
 
 func DeriveShaMux(list types.DerivableList, num *big.Int) common.Hash {
-	return instances[getType()].DeriveSha(list)
+	return instances[getType(num)].DeriveSha(list)
 }
 
 func EmptyRootHashMux(num *big.Int) common.Hash {
-	return emptyRoots[getType()]
+	return emptyRoots[getType(num)]
 }
 
-// TODO: Choose appropriate DeriveShaImpl from governance based on block number
-func getType() int {
+func getType(num *big.Int) int {
 	implType := config.DeriveShaImpl
+
+	// gov == nil if blockchain.InitDeriveSha() is used, in genesis block manipulation
+	// and unit tests. gov != nil if blockchain.InitDeriveShaWithGov is used,
+	// in ordinary blockchain processing.
+	if gov != nil {
+		if pset, err := gov.ParamsAt(num.Uint64()); err != nil {
+			logger.Crit("Cannot determine DeriveShaImpl", "num", num.Uint64(), "err", err)
+		} else {
+			implType = pset.DeriveShaImpl()
+		}
+	}
+
 	if _, ok := instances[implType]; ok {
 		return implType
 	} else {
```

Comment:
I meant this suggestion:
```





if _, ok := instances[implType]; ok {
    return implType
} else {
    //logger.Error("Unrecognized deriveShaImpl, falling back to Orig", "impl", implType)
    //return types.ImplDeriveShaOriginal
    logger.Error("Unrecognized deriveShaImpl, falling back to default impl", "impl", implType)
    return int(params.DefaultDeriveShaImpl)
}
```
Also, the type `params.DefaultDeriveShaImpl` may be changed to `int` as the impl types are already positioned as `int`.

Reply:
Fixed [3020f6f](https://github.com/klaytn/klaytn/pull/1708/commits/3020f6fd86290135a78503cb8c01f99caf296c04).

params.DefaultDeriveShaImpl is an uint64 to homogenize with existing integer parameters. (https://github.com/klaytn/klaytn/blob/v1.9.1/governance/default.go#L596)